Post Formats: Fix embeds in audio & video formats

ryelle opened the pull request:
See #391 — This iterates on the post formats code a little. First, it fixes an issue with embeds, and then it swaps the `excerpt-*` templates to use `has_block` to check for blocks, rather than having the conditional also be responsible for outputting the content.

Embed block issue: Once created, embed blocks appear to use `core-embed/spotify`, `core-embed/youtube` etc, so the detection of `core/embed` isn't working. This introduces a rough wildcard match. Now we can pass in `core-embed/*` to `twenty_twenty_one_print_first_instance_of_block`, and it'll return the first embed.

The `has_block` change was introduced to avoid confusion when looking at the template code. It's not immediately clear where the block content is printed out, since it's happening in the conditional. This change cleans up the conditionals so they're more understandable.

(Happy to split this PR in two for each change if anyone asks, I was already here so ended up doing both in one 😬 )

## Test instructions

This PR can be tested by following these steps:
1. Create an audio or video post using an embed
2. Add some extra content to the post, maybe add an excerpt
3. View your post in an archive, it should display the embed and the excerpt
4. View the post (singular view), it should show the full post content

---

1. Try creating other post formats
2. They should each display using the logic on #367

## Quality assurance
* [x] I have thought about any security implications this code might add.
* [x] I have checked that this code doesn't impact performance (greatly).
* [x] I have tested this code to the best of my abilities
* [x] I have checked that this code does not affect the accessibility negatively

ryelle committed:
Post Formats: Check for all embed blocks on audio & video posts

ryelle committed:
Use `has_block` instead of having the print function do double-duty

ryelle committed:
Prefer elseif for phpcs

aristath commented on template-parts/excerpt/excerpt-chat.php:
```diff
@@ -9,7 +9,6 @@
-}
+twenty_twenty_one_print_first_instance_of_block( 'core/paragraph', get_the_content(), 2 );
+
+the_excerpt();
```
The excerpt here should only be printed if we were unable to find paragraph blocks, so basically for old, non block-based content.

aristath commented on template-parts/excerpt/excerpt-video.php:
```diff
@@ -9,16 +9,13 @@
+$content = get_the_content();
 
-	// Fallback to the content if no block was found.
-	the_content();
```
If there is no `core/video` or `core/embed`/`core-embed/*` block we'll need to fallback to `the_content()` instead of `the_excerpt()`

ryelle replied:
@melchoyce & I just talked this through, and since these templates are only used when the "excerpt" view is shown, it doesn't make sense to show the full content in some cases. We should show the audio/video/embed if we can find it, and the excerpt always.

Based on that convo I've also updated the other `excerpt-*` templates, since only aside, status, and link should use `the_content`.

aristath replied:
This is a problem....
There are hundreds of sites that have older content, and the current approach will only work for block-based posts. That's why we had the `the_content()` fallbacks.
If the site has posts that are not block-based, then `the_excerpt` will not render anything.
You can test this for example by importing the test XML file from https://raw.githubusercontent.com/WPTT/theme-unit-test/master/themeunittestdata.wordpress.xml which contains old post-format posts.
Falling-back to `the_content` instead of `the_excerpt` will take into account those sites.

I added a PR against your fork with the tweaks I had in mind (https://github.com/ryelle/twentytwentyone/pull/2)
I apologize for the extra files in the diff, I pulled from trunk first

melchoyce replied:
> There are hundreds of sites that have older content, and the current approach will only work for block-based posts. That's why we had the the_content() fallbacks.

I'm okay with this. This theme should show off the best of new features and show off blocks. Older posts aren't broken — you can convert the older content to blocks if you want your site to take advantage of the block-formatted post format excerpts on archive pages. And if that's a problem, the option to show all of your content exists via the Customizer setting.

aristath replied:
This PR changes the syntax and makes an awesome job at making it more readable for developers. But it doesn't actually "fix" anything for users, in fact it breaks backwards-compatibility. 
Is improving code readability for developers worth a clear regression for end-users/visitors?

melchoyce replied:
This theme is designed to show off the block editor rather than the classic editor. If folks want to use the classic editor, there are previous default themes available that look great. If they want their site to support post formats with this theme, there are a few plugins I can recommend to convert old classic content to blocks.

aristath commented on template-parts/excerpt/excerpt-audio.php:
```diff
@@ -9,17 +9,13 @@
+$content = get_the_content();
 
-	// Fallback to the content.
-	the_content();
```
If there is no `core/audio` or `core/embed`/`core-embed/*` block we'll need to fallback to `the_content()` instead of `the_excerpt()`

ryelle committed:
Fix up the display of content vs excerpt

ryelle committed:
Update file comments

ryelle committed:
Bring back support for `core/embed` block

This block is currently only in GB, so we need to support both `core-embed/*` and `core/embed`

carolinan commented:
Screenshot of post formats with trunk:

![postformat](https://user-images.githubusercontent.com/7422055/96252316-0eb0df80-0fb2-11eb-8278-7e079abf907a.png)

With this PR:
![postformats-456](https://user-images.githubusercontent.com/7422055/96252501-5e8fa680-0fb2-11eb-88c6-0050836795b1.png)


-I prefer the trunk version that is not empty.
